File: src-tauri/src/commands/stability.rs
//! Tauri commands for stability reports and telemetry.

use crate::{
    models::stability::{CommandTelemetryEntry, StabilityReport},
    services::stability_service,
};

#[tauri::command]
pub fn stability_list_reports() -> Result<Vec<StabilityReport>, String> {
    stability_service::StabilityReport::list()
}

#[tauri::command]
pub fn stability_read_report(id: String) -> Result<StabilityReport, String> {
    stability_service::StabilityReport::read(&id)
}

#[tauri::command]
pub fn stability_delete_report(id: String) -> Result<(), String> {
    stability_service::StabilityReport::delete(&id)
}

#[tauri::command]
pub fn stability_mark_seen(id: String) -> Result<(), String> {
    stability_service::StabilityReport::mark_seen(&id)
}

#[tauri::command]
pub fn stability_unseen_count() -> Result<usize, String> {
    stability_service::StabilityReport::unseen_count()
}

#[tauri::command]
pub fn stability_recent_telemetry(
    limit: Option<u32>,
) -> Result<Vec<CommandTelemetryEntry>, String> {
    Ok(stability_service::recent_telemetry(
        limit.unwrap_or(50) as usize
    ))
}

#[tauri::command]
pub fn stability_violations() -> Result<Vec<CommandTelemetryEntry>, String> {
    Ok(stability_service::violations())
}

#[tauri::command]
pub fn stability_renderer_heartbeat() -> Result<(), String> {
    stability_service::renderer_heartbeat();
    Ok(())
}

/// Persist a renderer-side crash (React error boundary, window error, or
/// unhandled promise rejection) as a stability report so it survives the
/// recovery reload/restart and surfaces in the Debug panel alongside Rust
/// panics and freezes. Rust panics are already persisted by the panic hook,
/// so the frontend only calls this for renderer-origin failures.
#[tauri::command]
pub fn stability_record_renderer_crash(
    source: String,
    message: String,
    details: String,
) -> Result<(), String> {
    let summary = format!("Renderer crash ({source}): {message}");
    stability_service::StabilityReport::write("renderer", &summary, &details)?;
    Ok(())
}

File: src/lib/stability.ts
import { invoke } from "@tauri-apps/api/core";

// Stability report and telemetry lib wrappers.

export type StabilityReport = {
  id: string;
  kind: string;
  timestamp: number;
  summary: string;
  details: string;
  seen: boolean;
};

export type CommandTelemetryEntry = {
  command: string;
  durationMs: number;
  timestamp: number;
  violation: boolean;
};

export async function stabilityListReports(): Promise<StabilityReport[]> {
  return invoke<StabilityReport[]>("stability_list_reports");
}

export async function stabilityReadReport(id: string): Promise<StabilityReport> {
  return invoke<StabilityReport>("stability_read_report", { id });
}

export async function stabilityDeleteReport(id: string): Promise<void> {
  return invoke("stability_delete_report", { id });
}

export async function stabilityMarkSeen(id: string): Promise<void> {
  return invoke("stability_mark_seen", { id });
}

export async function stabilityUnseenCount(): Promise<number> {
  return invoke<number>("stability_unseen_count");
}

export async function stabilityRecentTelemetry(limit?: number): Promise<CommandTelemetryEntry[]> {
  return invoke<CommandTelemetryEntry[]>("stability_recent_telemetry", { limit });
}

export async function stabilityViolations(): Promise<CommandTelemetryEntry[]> {
  return invoke<CommandTelemetryEntry[]>("stability_violations");
}

export async function stabilityRendererHeartbeat(): Promise<void> {
  return invoke("stability_renderer_heartbeat");
}

/** Persist a renderer crash so it survives the recovery reload/restart and
 *  appears in the Debug panel alongside Rust panics and freezes. Best-effort;
 *  callers ignore failures so recovery is never blocked. */
export async function stabilityRecordRendererCrash(
  source: string,
  message: string,
  details: string,
): Promise<void> {
  return invoke("stability_record_renderer_crash", { source, message, details });
}
